### PR TITLE
Implemented route /v2/_catalog API functionality

### PR DESCRIPTION
**Pull request by keyskull**

### CODE REVIEW

**Comment by keyskull** [comment]
Anything else that I need to cover to merge this PR to the branch?

**Comment by skepticfx** [comment]
It looks there is no more user namespace or special authentication involved here. Shouldn't catalog be restricted to a privileged user?

**Comment by gabivlj** [comment]
> Anything else that I need to cover to merge this PR to the branch?

Sorry forgot to submit my review :-). There are some comments I wanted to be addressed

File: src/registry/r2.ts [hunk 1]
```diff
@@ -130,6 +131,33 @@ export class R2Registry implements Registry {
     return checkManifestResponse;
   }
 
+  async listRepositories(limit?: number): Promise<RegistryError | ListRepositoriesResponse> {
+    const env = this.env;
+    const options = {
+      limit: limit ? limit : 1000,
```
[the remark this answers is not in the document]
Comment (keyskull) [reply src/registry/r2.ts:137]: Sure

**Comment by keyskull** [comment]
About @skepticfx mentioned authentication feature, @gabivlj do you know we should have restrictions for each user or just general authentication?

File: src/registry/r2.ts [hunk 2]
```diff
@@ -131,11 +131,12 @@ export class R2Registry implements Registry {
     return checkManifestResponse;
   }
 
-  async listRepositories(limit?: number): Promise<RegistryError | ListRepositoriesResponse> {
+  async listRepositories(limit?: number, last?: string): Promise<RegistryError | ListRepositoriesResponse> {
     const env = this.env;
-    const options = {
+    const options:R2ListOptions = {
```
Comment (gabivlj) [inline src/registry/r2.ts:136]: nit: This is our bad due to not setting it up in the repo, but could we have the file linted with prettier?

**Comment by gabivlj** [comment]
Thank you @keyskull, I had another nit but once that is addressed I think all is ok to merge from my side.

**Comment by gabivlj** [comment]
> About @skepticfx mentioned authentication feature, @gabivlj do you know we should have restrictions for each user or just general authentication?

We have some kind of read/write permissions. IMO this is already handled on the middleware side.

File: src/registry/r2.ts (hunk 2 above)
Comment (keyskull) [reply src/registry/r2.ts:136]: Sure!

**Comment by skepticfx** [comment]
@gabivlj registries don't allow access to this by default AFAIK. But I believe, this specific authorization check is not really relevant to this PR. Believe a middleware which does the right RBAC should handle this correctly. None of our middleware examples allow this today, but something to be cautious about.

**Review by skepticfx — APPROVED** [review]

**Pull request merged**